File: monotonicStack/3.sumOfSubarrayMinimums/js/sumOfSubarrayMinimums.js

/**
 * @param {number[]} arr
 * @return {number}
 */
const MOD = 1000000007
var sumSubarrayMins = function(arr) {
    const stack = []
    const n = arr.length
    let ans = 0
    for(let i in arr){
        while(stack.length > 0 && arr[stack.at(-1)] >= arr[i]){
            const top = stack.pop()
            // 以当前值为最小值的子数组有多少个
            // 左边界
            const left = stack.length == 0 ?-1:stack.at(-1)
            ans = (ans + (top-left)*(i-top)*arr[top]) % MOD
        }
				stack.push(i)
    }
    // 清理栈中剩余的值
    while(stack.length > 0){
        const top = stack.pop()
        const left = stack.length == 0 ?-1:stack.at(-1)
        ans = (ans + (top-left)*(n-top)*arr[top]) % MOD
    }
    return ans
};

File: monotonicStack/3.sumOfSubarrayMinimums/go/sumOfSubarrayMinimums.go
package _go

const MOD = 1000000007

var stack = [100000]int{}
var r int

func sumSubarrayMins(arr []int) int {
	ans := 0
	r = 0
	n := len(arr)
	for i := 0; i < n; i++ {
		for r > 0 && arr[stack[r-1]] >= arr[i] {
			top := stack[r-1]
			r--
			// 判断子数组的个数
			left := -1
			if r > 0 {
				left = stack[r-1]
			}
			ans = (ans + (i-top)*(top-left)*arr[top]) % MOD

		}
		stack[r] = i
		r++
	}
	// 栈中剩余的值
	for r > 0 {
		top := stack[r-1]
		r--
		// 判断子数组的个数
		left := -1
		if r > 0 {
			left = stack[r-1]
		}
		ans = (ans + (n-top)*(top-left)*arr[top]) % MOD
	}
	return ans
}
